src/serializers/serializer_io.c

```diff
@@ -168,20 +178,282 @@ char *SerializerIO_ReadBuffer
+                                                                \
+	/* ensure there's at least sizeof(t) bytes in buffer */     \
+	ASSERT((buffer->cap - buffer->count) >= REQUIRED_SIZE(t));  \
+                                                                \
+	/* validate type */                                         \
+	DEBUG_VALIDATE_TYPE(t)                                      \
+                                                                \
+	/* read value */                                            \
+	t v = *(t*)(buffer->buffer + buffer->count);                \
+                                                                \
+	/* update offset */                                         \
+	buffer->count += sizeof(t);                                 \
+                                                                \
+	return v;                                                   \
+}
+
+//------------------------------------------------------------------------------
+// create buffer serializer read & write functions
+//------------------------------------------------------------------------------
+
+// BufferSerializerIO_ReadFloat & BufferSerializerIO_WriteFloat
+BUFFERED_SERIALIZER_READ_WRITE(Float, float)
+
+// BufferSerializerIO_ReadDouble & BufferSerializerIO_WriteDouble
+BUFFERED_SERIALIZER_READ_WRITE(Double, double)
+
+// BufferSerializerIO_ReadSigned & BufferSerializerIO_WriteSigned
+BUFFERED_SERIALIZER_READ_WRITE(Signed, int64_t)
+
+// BufferSerializerIO_ReadUnsigned & BufferSerializerIO_WriteUnsigned
+BUFFERED_SERIALIZER_READ_WRITE(Unsigned, uint64_t)
+
+// BufferSerializerIO_ReadLongDouble & BufferSerializerIO_WriteLongDouble
+BUFFERED_SERIALIZER_READ_WRITE(LongDouble, long double)
+
+// write buffer to stream
+void BufferSerializerIO_WriteBuffer
+(
+	void *io,           // serializer
+	const char *value,  // value
+	size_t len          // value size
+) {
+	ASSERT(io    != NULL);
+	ASSERT(value != NULL);
+	ASSERT(len   > 0);
+
+	BufferedIO *buffer = (BufferedIO*)io;
+
+	// make sure value has enough room
+	if(_accommodate(buffer, len + REQUIRED_SIZE(size_t))) {
+		// in DEBUG mode we write the value type
+		DEBUG_WRITE_TYPE(char*)
+
+		// add to buffer
+		// write value length to stream
+		*((size_t*)(buffer->buffer + buffer->count)) = len;
+		buffer->count += sizeof(size_t);
+
+		// write value to buffer
+		memcpy(buffer->buffer + buffer->count, value, len);
+		buffer->count += len;
+	} else {
+		// value is too big
+		// buffer had been flushed by '_accommodate'
+		RedisModule_SaveStringBuffer(buffer->stream, value, len);
+	}
+}
+
+// read buffer from stream
+char *BufferSerializerIO_ReadBuffer
+(
+	void *io,       // stream
+	size_t *lenptr  // number of bytes to read
+) {
+	BufferedIO *buffer = (BufferedIO*)io;
+
+	// load buffer if depleted
+	if(unlikely(buffer->count == buffer->cap)) {
+		_load_buffer(buffer);
+	}
+
+	// check for large string
+	if(unlikely(buffer->cap > BUFFER_SIZE)) {
```

Comment:
can't we use the same style in BufferSerializerIO_WriteBuffer and check the _load_buffer return boolean?